Supply chain and delivery can set serviceAccountRef

martyspiewak opened the pull request:
Closes #301

Commit:
Supply chain and delivery can set serviceAccountRef

waciumawanjohi commented:
This needs documentation.
At least in the [reference doc](https://github.com/vmware-tanzu/cartographer/blob/main/site/content/docs/development/reference.md)

(I don't think the serviceAccount in the workload has been documented there, this would be a good time to add it)

martyspiewak commented:
> This needs documentation. At least in the [reference doc](https://github.com/vmware-tanzu/cartographer/blob/main/site/content/docs/development/reference.md)
> 
> (I don't think the serviceAccount in the workload has been documented there, this would be a good time to add it)

Makes sense. I added documentation for both in the latest commit.

waciumawanjohi commented on pkg/controller/deliverable/reconciler_test.go:
```diff
@@ -261,19 +261,50 @@ var _ = Describe("Reconciler", func() {
 			Expect(resourceRealizer).To(Equal(builtResourceRealizer))
```
I was wondering why this assert wasn't in the context of the service account specified in the delivery or when using the default service account. Should this be a separate "It" block?

martyspiewak replied:
This makes sense to me. Addressed in latest commit.

waciumawanjohi commented on pkg/registrar/map_functions_test.go:
```diff
@@ -1248,16 +1278,205 @@ var _ = Describe("MapFunctions", func() {
+								"some-label": "some-label-value",
+							},
+						},
+						Spec: v1alpha1.WorkloadSpec{
+							ServiceAccountName: "some-service-account",
+						},
```
There are 50 lines of boilerplate in the BeforeEach, but the only thing that should vary between the last test and this one is this. It's difficult to verify that the rest of the boilerplate is the same. That's why I advocate for deduplicating tests. Following how objects are built up from parent contexts is much easier than manually comparing dozens of lines of code for changes.

martyspiewak replied:
I completely agree and I began to refactor the tests to do exactly that but after discussing with others we determined that we would like to refactor the whole map_functions.go and so it didn't seem worth it to refactor the tests right now if we will (hopefully) soon begin another refactor.

martyspiewak commented:
Squashed and rebased on main and force-pushed. 

Re-requesting a review because I switched from using `serviceAccountName` to `serviceAccountRef` per the discussion in #301

idoru commented on pkg/apis/v1alpha1/cluster_supply_chain_test.go:
```diff
@@ -55,6 +55,14 @@ var _ = Describe("ClusterSupplyChain", func() {
 			Expect(jsonValue).NotTo(ContainSubstring("omitempty"))
 		})
+
+		It("allows but does not require a service account name", func() {
```
s/name/ref/

idoru commented on pkg/apis/v1alpha1/common.go:
```diff
@@ -84,6 +84,11 @@ type ObjectReference struct {
 	APIVersion string `json:"apiVersion,omitempty"`
 }
 
+type NameAndNamespace struct {
```
I think it would be more intent revealing if it were named `ServiceAccountRef[erence]`

martyspiewak replied:
I'm definitely open to renaming that. I think it might make sense to keep it generic because the only fields on there are `name` and `namespace` which are not specific to service accounts.

We already have an `ObjectReference` type which has kind, apiversion, name, namespace. My first thought was we could call this new type `SimpleObjectReference` or something to that effect.

What do you think?

idoru replied:
The reason I brought this up was that I was drawing parallels with `ClusterTemplateReference` and `DeliveryClusterTemplateReference`. Additionally, I considered design of core k8s types, say `ClusterRoleBinding`  -- you'll see the role binding has a very specific type for the ref `RoleRef` even though this has very generic fields like `APIGroup` , `Kind` and `Name`.

martyspiewak replied:
Interesting. I think the ClusterTemplateRef is different because we’re doing validation on those to ensure that only certain kinds are used. 

But the RoleRef seems similar enough that I’m convinced we can change to ServiceAccountRef. If we ever have a need to reuse the type we can think about renaming then

idoru commented on site/content/docs/development/reference.md:
```diff
@@ -148,6 +154,12 @@ spec:
+  # is not specified in the workload
+  #
+  # (optional)
+  serviceAccountName: service-account
```
Should be `serviceAccountRef` instead now?

idoru reviewed (commented):
See individual comments. Summary:
- prefer more explicit naming for `serviceAccountRef`'s type
- test prose (name -> ref)
- docs for supplychain should show `serviceAccountRef` not `serviceAccountName`

Commit:
Address reviewers comments in docs and test name

Commit:
Address reviewers comments about type name
